lib/ansible/plugins/inventory/docker_machine.py

```diff
@@ -0,0 +1,159 @@
+        - The plugin returns an I(all) group of nodes and one group per driver (e.g. digitalocean).
+        - The plugin sets standard host variables I(ansible_host), I(ansible_port), I(ansible_user) and I(ansible_ssh_private_key).
+        - The plugin also sets standard host variable I(ansible_ssh_common_args) to '-o StrictHostKeyChecking=no'.
+        - The plugin also stores the Docker Machine 'env' variables in I(dm_) prefixed host variables.
+
+    options:
+        plugin:
+          description: token that ensures this is a source file for the C(docker_machine) plugin.
+          required: True
+          choices: ['docker_machine']
+        verbose_output:
+            description: Toggle to (not) include all available nodes metadata (e.g. Image, Region, Size) as a JSON object.
+            type: bool
+            default: yes
+        split_tags:
+          description: for keyed_groups add two variables as if the tag were actually a key value pair separated by a colon, instead of just a single value.
+          required: False
+          type: bool
+          default: no
+        split_separator:
+          description: for keyed_groups when splitting tags this is the separator to split the tag value on.
+          required: False
+          type: str
+          default: ":"
+'''
+
+EXAMPLES = '''
+# Minimal example
+plugin: docker_machine
+
+# Example using constructed features to create groups
+# keyed_groups may be used to create custom groups
+strict: False
+keyed_groups:
+  - prefix: tag
+    key: 'dm_tags'
+
+# Example using tag splitting where the tag is like 'dm_tag_gantry_component:routinator'
+strict: False
+split_tags: True
+split_separator: ":"
+keyed_groups:
+  - prefix: gantry_component
+    key: 'dm_tag_gantry_component'
+'''
+
+from ansible.errors import AnsibleError
+from ansible.module_utils._text import to_native
+from ansible.module_utils._text import to_text
+from ansible.plugins.inventory import BaseInventoryPlugin, Constructable, Cacheable
+
+import json
+import re
+import subprocess
+
+
+class InventoryModule(BaseInventoryPlugin, Constructable, Cacheable):
+    ''' Host inventory parser for ansible using Docker machine as source. '''
+
+    NAME = 'docker_machine'
+
+    def _run_command(self, *args):
+        return subprocess.check_output(["docker-machine"] + list(args)).decode('utf-8').strip()
+
+    def _populate(self):
+        self.inventory.add_group('all')
+
+        try:
+            self.nodes = self._run_command('ls', '-q').splitlines()
+            for self.node in self.nodes:
+                self.node_attrs = json.loads(self._run_command('inspect', self.node))
+
+                id = self.node_attrs['Driver']['MachineName']
+                self.inventory.add_host(id)
+                self.inventory.add_group(self.node_attrs['DriverName'])
+                self.inventory.add_host(id, group=self.node_attrs['DriverName'])
```

Comment:
Why not use keyed groups and let the users decide themselves whether they want to create such a group, instead of creating it by default? (There should be an example of how to do this if the user is suposed to do it by herself.)

Reply:
No idea, I'm fairly new to Ansible so I have a very limited understanding of these features.

Reply:
I'm pretty new to inventory plugins (and never used them personally, except for tests), but that is a pretty cool feature which allows users to add all kind of groups. Probably something like
```
keyed_groups:
    # this creates a group per driver (...some examples...) and assigns the hosts that have matching values to it
    - separator: ''
      key: docker_machine_node_attributes.DriverName
```
should do the trick.

Reply:
Ah that's excellent, I have removed the explicit group creation from the code and added an example instead based on your example fragment.